commit 78c2274fad7c6bf14e7067ce941f95009e3eda4f
Author: Petr Hosek <phosek@chromium.org>
Date:   Mon Sep 10 21:54:56 2018 +0000

    [ELF] Don't emit .relr.dyn section if there are no relocs
    
    This resolves PR38875.
    
    Differential Revision: https://reviews.llvm.org/D51834
    
    git-svn-id: https://llvm.org/svn/llvm-project/lld/trunk@341870 91177308-0d34-0410-b5e6-96231b3b80d8

diff --git a/ELF/SyntheticSections.h b/ELF/SyntheticSections.h
index a14e831f0..f51e85c17 100644
--- a/ELF/SyntheticSections.h
+++ b/ELF/SyntheticSections.h
@@ -536,6 +536,7 @@ struct RelativeReloc {
 class RelrBaseSection : public SyntheticSection {
 public:
   RelrBaseSection();
+  bool empty() const override { return Relocs.empty(); }
   std::vector<RelativeReloc> Relocs;
 };
 
diff --git a/test/ELF/empty-pack-dyn-relocs.s b/test/ELF/empty-pack-dyn-relocs.s
new file mode 100644
index 000000000..b74a89d66
--- /dev/null
+++ b/test/ELF/empty-pack-dyn-relocs.s
@@ -0,0 +1,11 @@
+// REQUIRES: x86
+
+// RUN: llvm-mc -filetype=obj -triple=x86_64-unknown-linux %s -o %t.o
+// RUN: ld.lld -pie --pack-dyn-relocs=relr %t.o -o %t
+// RUN: llvm-readobj -sections %t | FileCheck %s
+
+.global _start
+_start:
+  nop
+
+# CHECK-NOT: Name: .relr.dyn
